rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,292 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 19, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      group-by src |
+      summarize count(dst) |
+      write feather |
+      to /path/to/file.feather'
+```
+
+Just to illustrate the composability, we could take this to an extreme when
+performance doesn't matter:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast group-by src |
+  vast 'summarize count(dst)' |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+The UX would translate seamlessly to other languages, e.g., Python:
+
+```python
+import vast as v
+await v.from("s3://aws")
+       .read_json() # or read("json")
+       .group_by("src")
+       .summarize(v.count("dst"))
+       .write_feather() # or write("feather")
+       .to("/path/to/file.feather")
+```
+
+### Pipeline Execution
+
+Let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
+`vast import zeek` as:
+
+```bash
+vast exec 'from - | parse zeek | to vast://1.2.3.4'
+```
+
+Likewise, `vast export json EXPRESSION` would become:
+
+```bash
+vast exec 'from vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The beauty is that we can almost write the same without requiring a VAST node
+and execute *in situ*:
+
+```bash
+vast exec 'from - |
+           where EXPRESSION |
+           write json |
+           to -'
```

Comment:
Good find. That's indeed an inconsistency. Your intuition is correct.

> `from vast://1.2.3.4` is maybe not very intuitive because the output is `Arrow`

Yes, I would propose to ditch `from vast://1.2.3.4`, but not `from` per se. I'm not sure about the replacement. If we go the route of "everything is an operator", then `vast "pull | where EXPR | ...` could make sense. If we want to have top-level commands, then your proposal of `vast exec "pull ..."` is better suited.

Intuitively, I'd go for the latter as of right now.

Reply:
@rdettai please resolve if you think this is captured in the latest version.